docs/test_requirements.md

```diff
@@ -0,0 +1,131 @@
+   - every valid URL pattern
+   - every response code
+   - every exception type
+      - service error
+      - service timeout
+      - resource does not exist
+      - insufficient permissions
+      - PubSub rejects individual message
+      - ingestion-edge:
+         - disk full after adding messages to queue
+         - disk already full
+         - corrupted SQLite database
+   - with gzipped and non-gzipped payload
+   - present, absent, and invalid required attributes
+   - present, absent, and invalid optional attributes
+   - present, absent, invalid, and wrong-type payload JSON
+   - present, absent, invalid, and wrong-type payload fields
+   - duplicate documents
+   - ingestion-beam: inputs and outputs return same schema received
+
+### Integration Tests
+
+Integration tests should have the following properties:
+
+- suite runs in less than 1 minute in CI
+- run against remotely configured service
+- use emulators or actual external services
+- cover all production behavior
+- cover production configuration options
+- cover all API endpoints
+- cover behavior with healthy external resources
+   - ingestion-edge: require disk queue not used
+- cover behavior with invalid traffic
+- cover behavior with exceptions from external services
+   - external service DNS does not resolve
+   - external service cannot TCP connect
+   - external service hangs after TCP connect
+   - external service returns 500
+   - external service resource does not exist
+   - external service insufficient permissions
+   - PubSub rejects individual messages
+   - ingestion-edge: all PubSub exceptions with queue disk full
+   - ingestion-beam pre-production: all exceptions that can be forcibly
+     produced with the Dataflow runner, which may include:
+      - PubSub returns 500
+      - PubSub cannot TCP connect
+      - Cloud Storage cannot TCP connect
+      - Dataflow disk is full
+- clean up external resources when complete
+
+### Load Tests
+
+Load tests should have the following properties:
+
+- can run in parallel in less than 1 hour
+- require no manual intervention to evaluate
+- evaluate performance against a tunable threshold
+- cover behavior with sustained 1.5x expected peak production traffic volume
+   - as of 2018-08-01 peak production traffic volume is about 11K req/s
+   - before, during, and after a period of:
+      - service downtime, except ingestion-edge
+      - 100%, 50%, and 5% invalid traffic
+      - external services unreachable
+      - 100% and 10% of external service requests time out
+      - 100% and 10% external service requests return 500
+   - ingestion-edge:
+      - when running out of disk queue space due to PubSub server errors
+      - require disk queue is not touched in absence of PubSub errors
+- full recovery from downtime must not take longer than the downtime period
+
+### Recovery Tests
+
+Recovery tests should the following properties in addition to load test
+properties:
+
+- do not need to run in 1 hour
```

Comment:
So this means they _may_ run for longer than 1 hour?

Reply:
It might be useful to adopt [RFC requirement levels](https://www.ietf.org/rfc/rfc2119.txt) for this document, so each line begin like "MUST x", "MAY y", "SHOULD NOT z"